### ISSUE TITLE
Make it easier to override toCloudFormation by making it recursive

### ISSUE THREAD

**Issue by rix0rrr** [awslabs/aws-cdk#606]
Right now, the `App` crawls the entire construct tree, calls `toCloudFormation` where it exists and aggregates the result.

This lack of implementation hiding gives users no ability to take existing constructs but tweak them slightly, as the only CloudFormation generation happens at the very lowest level. As came up in https://github.com/awslabs/aws-cdk/issues/603, since every L2 is already wrapping another resource, in this very simplest construct tree:

```
ClassicLoadBalancer
  |
  +--- AWS::ELB::LoadBalancer
```

Let's say a user wanted to take the output of the current `ClassicLoadBalancer` and mostly keep it but add an attribute.

In a normal OO system, you would subclass `ClassicLoadBalancer`, override `toCloudFormation` and change the parent's output slightly before returning it.

In our system, you can't do this since `ClassicLoadBalancer` is not involved in the synthesis process at all, and consumers have no control over the instance of `AWS::ELB::LoadBalancer` that is constructed INSIDE the `ClassicLoadBalancer`.

I would argue it would be beneficial if we could achieve the former, since it conforms to expectations and it also provides a more usable "escape hatch" mechanism (that customers are asking for) than simply overriding `Stack.toCloudFormation()`.

A simple way of achieving this would be to have `App.toCloudFormation()` stop recursing into children as soon as it finds a construct that implements `toCloudFormation()`, and leave it up to the highest-level resource in the tree to recurse further and synthesize all of its children (optionally twiddling the output). This makes it so that implementing `ClassicLoadBalancer.toCloudFormation()` actually has the opportunity to collect all of its children, modify the output, then return that.

@eladb, opinions?

**Comment by eladb** [comment]
@rix0rrr proposed:

Right now, the `App` crawls the entire construct tree, calls `toCloudFormation` where it exists and aggregates the result.

This lack of implementation hiding gives users no ability to take existing constructs but tweak them slightly, as the only CloudFormation generation happens at the very lowest level. As came up in #603, since every L2 is already wrapping another resource, in this very simplest construct tree:

```
ClassicLoadBalancer
  |
  +--- AWS::ELB::LoadBalancer
```

Let's say a user wanted to take the output of the current `ClassicLoadBalancer` and mostly keep it but add an attribute.

In a normal OO system, you would subclass `ClassicLoadBalancer`, override `toCloudFormation` and change the parent's output slightly before returning it.

In our system, you can't do this since `ClassicLoadBalancer` is not involved in the synthesis process at all, and consumers have no control over the instance of `AWS::ELB::LoadBalancer` that is constructed INSIDE the `ClassicLoadBalancer`.

I would argue it would be beneficial if we could achieve the former, since it conforms to expectations and it also provides a more usable "escape hatch" mechanism (that customers are asking for) than simply overriding `Stack.toCloudFormation()`.

A simple way of achieving this would be to have `App.toCloudFormation()` stop recursing into children as soon as it finds a construct that implements `toCloudFormation()`, and leave it up to the highest-level resource in the tree to recurse further and synthesize all of its children (optionally twiddling the output). This makes it so that implementing `ClassicLoadBalancer.toCloudFormation()` actually has the opportunity to collect all of its children, modify the output, then return that.

@eladb, opinions?

**Comment by rix0rrr** [comment]
I thiiiink this might be part of the appmodel refactor?

**Comment by rix0rrr** [comment]
The synthesis implementation we use right now looks somewhat like this:

```ts
class Stack extends Construct {
    public toCloudFormation() {
        walk(this.children, c => {
            if (c instanceof StackElement) {
                return c.toCloudFormation();
            }
        });
    }
}
```
I.e., there's a single class that walks the entire tree, and for every child checks if it looks like something that can produce CloudFormation and if so calls it.

The fact that a single piece of code walks the entire tree and calls the producers, without giving intervening classes the ability to intercept this, is exactly what's making the "escape hatch" problem hard, and is making it so that people have to override `Stack.toCloudFormation()`.

What I'm arguing is that the synthesis implementation should always have been like this:

```ts
class Construct {
    public toCloudFormation() {
        return merge(this.children.map(c => c.toCloudFormation());
    }
}
```

So that every Construct in the tree has the ability to modify the returned values from their children, and we have a natural escape hatch mechanism.

We couldn't do this before, because we wanted `Construct`s to be more general than CloudFormation, but we COULD do this now by abstracting away from CloudFormation, and calling it `toAppModel`:

```ts
class Construct {
    public toAppModel() {
        return merge(this.children.map(c => c.toAppModel());
    }
}
```

**Comment by rix0rrr** [comment]
For example, I think CloudFormation resources can emit the following:

```
const SomeResource {
     public toAppModel() {
        return {
            CloudFormationResource {
                [this:logicalId]: {
                    Type: 'AWS::IAM::Oei'
                }
            }
        };
    }
}
```

And then `Stack` can collect up all keys under `CloudFormationResource` and emit a `CloudFormationStack`. Anything that's not recognized at the end is an error.

**Comment by ccurrie-amzn** [comment]
Some thoughts:
* As I user, I want to use the L1 constructs for my "escape hatch", not manipulate the generated CF template. L1 constructs have an API similar to the L2 and "feel" more natural to work with having learned the L2 API. Other factors:
  * I have a need to set dependencies between L1 and L2 constructs; while I can see how this would be possible at "conversion" time, it feels more natural to do at the construct level.
  * I don't want to have to create a derived class just to set an unsupported property or specify an extra dependency; the extra boilerplate will quickly get tiresome.
  * But I would still prefer to specify my modifications in the context where the construct being modified is declared, rather than having them in a separate "conversion" method.

**Comment by eladb** [comment]
@rix0rrr updated based on your inputs.
@RomainMuller this is what I meant by "Overrides should be applied on top of anything L2 does". I will clarify

**Comment by RomainMuller** [comment]
Well - some of the validation happens at the L1 level...

**Comment by eladb** [comment]
@RomainMuller good point. So you are suggesting that any "patches" should also bypass any L1 validations, in case the spec is wrong or not up-to-date

**Comment by RomainMuller** [comment]
A caveat I see as well is that if the "primary" resource of an L2 changes (aka used to be type `Foo` and now is `Bar`), the down-cast will still compile (because you're forcing the type checker's hand anyway), but you might get surprising results (hopefully, that don't synthesize or deploy, but if you're merely changing resource `options`, then it will happily proceed).

Also, are we positive the find & down-case solution can be used in all the `jsii` front-end languages?

**Comment by eladb** [comment]
@kiiadi good point, this is actually possible (`instanceof` is usually not a good practice in JavaScript, but one can use `resourceType`):

```ts
bucket.children.find(c => (c as cdk.Resource).resourceType === 'AWS::S3::Bucket') 
    as s3.cloudformation.BucketResource;
```

@RomainMuller down-casting is a basic object-oriented capability and I want to believe it's going to be supported by all the languages we bind to. In strongly-typed languages like Java, if the actual resource object has a different type, the down cast will fail at runtime. In JavaScript, we might get some weird behavior, for sure. Can you elaborate on your concern with interface proxies? Example?

**Comment by eladb** [comment]
Names:

| 1 | addOverride(p, v) | addPropertyOverride(pp, v) | propertyOverrides.X
|---|---------------|-------------------------|------------------------------------
| 2 | override(p,v) | overrideProperty(p,v) | propertyOverrides.X
| 3 | patch(p,v) | patchProperty(pp,v) | overrides.X |
| 4 | | setProperty(pp,v) | bucketOverrides.X |

I am leaning towards: `patch`, `patchProperty` and `propertyOverrides`. Thoughts?

**Comment by RomainMuller** [comment]
@eladb  - Regarding the interface proxies, I reckon it's never an issue when dealing with `Resource` classes (since they are... classes, y'know!).

Regarding the name proposals... To me, `patch` implies you're going to provide it in the for of `add`, `remove` and `replace` operations, and your examples seem to imply you're not going to get them this way... So I would vote against `patch` specifically here.

The `setProperty` option also looks wrong to me, because it looks way too "normal" (and you probably want those overrides to not look like normal, instead to look like "here be dragons 🐲" / "I know what I'm doing here").

**Comment by rix0rrr** [comment]
Ha, let me be contrarian. I like `patch` because it's very clear what you're doing: fixing up some deficiency.

Also, I generally prefer shorter words to longer words for verbs in programming, therefore `patch` (1 syllable) is better than `override` (3 syllables).

**Comment by ccurrie-amzn** [comment]
> This means that in order to delete a value, you will have to set it to null in the overrides.

Null is a valid and reasonable value for some properties; I've run into issue trying to replicate what the console does in CDK and been blocked from using `null` where it appeared in my console test output. I would instead like to see an explicit property removal API.

**Comment by eladb** [comment]
Good point about property removal... 

Regardless, curious about your use case for using `null` in CloudFormation property values. Can you point me to an example?

**Comment by ccurrie-amzn** [comment]
It occurred in APIGateway, when I used the console to enable CORS. When the result was examined using `aws apigateway get-method`, `methodIntegration.integrationResponses` had

```
responseTemplates: {
    "application/json": null
}
```

You can see this in the documentation as well: https://docs.aws.amazon.com/cli/latest/reference/apigateway/get-method.html

I'm not able to replicate this in CDK, because `apigateway.generated.d.ts` defines `IntegrationResponseProperty.responseTemplates` as non-nullable.

**Comment by eladb** [comment]
@ccurrie-amzn updated requirements and design (now moved to issue description) to reflect the requirement to allow `null`s and deletion of values. Thoughts/

**Comment by ccurrie-amzn** [comment]
I think it looks good enough for getting started. Hands on experience will be more instructive at this point. 😄

**Comment by moofish32** [comment]
I agree with @ccurrie-amzn, on paper my pain points look covered by the content in the description.

**Comment by eladb** [comment]
Thanks for your inputs guys! Looking forward for your feedback from actual use soon!

**Comment by eladb** [comment]
About when to apply L1 schema validation...

We are going to support two layers of overrides: 

1. Strongly-typed property structure, e.g.

```ts
bucketResource.propertyOverrides.corsConfiguration = {
    corsRules: [
        {
            allowedMethods: [ 'GET' ],
            allowedOrigins: [ '*' ]
        }
    ]
};
```

2. Weakly-typed "patches", e.g.

```ts
bucketResource.addPropertyOverride('VersioningConfiguration.Status', 'NewStatus');
```

The second option will allow _anything_ to be assigned, and will bypass all L1 schema validations. The question is: do you think it makes sense for the first type of overrides to go through schema validations or not. 

Initially we were saying that it should bypass schema validation and allow anything to be assigned, but when diving a little deeper, some of the schema validation actually happens at compile time (e.g. requires/optional props), so does it make sense to opt-out of some of the validation in this usage?

For the cases where the CFN schema is wrong (and thus requires an override), users could just go to the 2nd option (`addPropertyOverride`) and opt-out of any type/schema checking.

**Comment by ccurrie-amzn** [comment]
Personally I like validation where possible. I think the default should be to validate the typed overrides, but I can see having the option to turn it off if you need to. You might consider tracking how often opt-outs happen in your usage metadata.

**Comment by eladb** [comment]
@ccurrie-amzn great ideas!

- Tracking overrides via CDK::Metadata might be a great way to prioritize our L2 work.
- Turning off validation at the resource level 👍 - we can do it regardless of overrides (another escape hatch mechanism)... Adding to requirements...

**Comment by moofish32** [comment]
I like the idea of tracking overrides in Metadata to target gaps. I probably lean more towards validation always for the typed properties and if that fails users have to go to option 2. We hopefully need very few use cases of option 2.

**Comment by eladb** [comment]
@ccurrie-amzn see #782 regarding "turning off" validation. TL;DR, it's non-trivial, so I am punting until we have evidence that it is actually needed/valuable.

**Comment by 0xdevalias** [comment]
Are these escape hatches officially/nicely documented anywhere on the main docs site? I stumbled across this issue by accident and wasn't aware of it prior, but it's a super useful feature that should be easily discoverable!

**Comment by eladb** [comment]
Here: https://awslabs.github.io/aws-cdk/aws-construct-lib.html#access-the-aws-cloudformation-layer

**Comment by 0xdevalias** [comment]
Must have missed that previously. Thanks!

**Comment by Console32** [comment]
@eladb is there a new version of the document? The old link is dead.

**Comment by DavidChristiansen** [comment]
@Console32 - here you go - https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html
